ftl/common/build_layer.py

```diff
@@ -0,0 +1,155 @@
+# Copyright 2017 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""This package defines the shared cli args for ftl binaries."""
+
+import abc
+import logging
+import datetime
+
+from containerregistry.client.v2_2 import docker_digest
+
+from ftl.common import ftl_util
+from ftl.common import tar_to_dockerimage
+
+_DEFAULT_TTL_WEEKS = 1
+
+class BaseLayerImage(object):
+    """BaseLayer is an abstract base class representing a container layer.
+
+    It provides methods for generating a dependency layer and an application
+    layer.
+    """
+
+    __metaclass__ = abc.ABCMeta  # For enforcing that methods are overriden.
+
+    def __init__(self):
+        self._img = None
+        self._lyr = None
+        self._diff_id = None
+        self._overrides = None
+
+    @abc.abstractmethod
+    def GetImage(self):
+        """Get the.
+        Returns:
+          a raw string of the layer's .tar.gz
+        """
+
+    @abc.abstractmethod
+    def GetLayer(self):
+        """Get the.
+        Returns:
+          a raw string of the layer's .tar.gz
+        """
+
+    @abc.abstractmethod
+    def GetDiffId(self):
+        """Get the.
+        Returns:
+          a raw string of the layer's .tar.gz
+        """
+
+    @abc.abstractmethod
+    def GetCacheKeyRaw(self):
+        """Synthesizes the application layer from the context.
+        Returns:
+          a raw string of the layer's .tar.gz
+        """
+
+    @abc.abstractmethod
+    def GetCacheKey(self):
+        """Synthesizes the application layer from the context.
+        Returns:
+          a raw string of the layer's .tar.gz
+        """
+
+    @abc.abstractmethod
+    def GetOverrides(self):
+        """Synthesizes the application layer from the context.
+        Returns:
+          a raw string of the layer's .tar.gz
+        """
+
+    @abc.abstractmethod
+    def BuildLayer(self):
+        """Synthesizes the application layer from the context.
+        Returns:
+          a raw string of the layer's .tar.gz
+        """
+
+# TODO(aaron-prindle) create custom error type
+
+class CacheLayerImage(BaseLayerImage):
+
+    def GetCacheRaw(self):
+        pass
+
+    def BuildLayer(self):
+        pass
+
+    def GetCacheKey(self):
+        return docker_digest.SHA256(self.GetCacheKeyRaw())
```

Comment:
Not sure we need this level of indirection - one method to GetCacheKey should probably suffice.

Reply:
This was done to make the aggregate/dependent keys more understandable.  Does it make more sense to have the combined key used hash values?